### ISSUE TITLE
spacy-langdetect example should be updated to work with more recent spaCy versions

### ISSUE THREAD

**Issue by benjlis** [explosion/spaCy#11038]
The spacy-langdetect example should be updated to work with more recent spaCy versions.

When I attempt to run the current documentation example under spaCy v3.3, I get this error:
ValueError: [E966] `nlp.add_pipe` now takes the string name of the registered component factory, not a callable component.

This error appears as an issue in the spacy-langdetect GitHub repo, along with updated code using modern spaCy:
`https://github.com/Abhijit-2592/spacy-langdetect/issues/6
`
Here is the code, and it works for me: 

```python
import spacy
from spacy.language import Language
from spacy_langdetect import LanguageDetector

@Language.factory("language_detector")
def create_language_detector(nlp, name):
    return LanguageDetector(language_detection_function=None)

nlp = spacy.load("en_core_web_sm")

nlp.add_pipe('language_detector')
text = 'This is an english text.'
doc = nlp(text)
# document level language detection. Think of it like average language of the document!
print(doc._.language)
# sentence level language detection
for sent in doc.sents:
   print(sent, sent._.language)
```

## Which page or section is this issue related to?
https://spacy.io/universe/project/spacy-langdetect

**Comment by polm** [comment]
Thanks for bringing our attention to this. Note that `spacy-lang-detect` is a third-party library, not developed by us, so we have no control over their README or code. We do control the Universe entry, though we've never proactively updated those before - they're posted and updated by library maintainers. In this case, since the upstream library seems abandoned, it seems more likely we'd remove it. 

My personal take is that language detection doesn't typically make sense in a spaCy pipeline - you'd usually want to do it on raw text as a preprocessing step, and then pass the sorted text into a spaCy pipeline corresponding the the appropriate language.

**Comment by benjlis** [comment]
Thanks for your advice. I'm new to spaCy and NLP, so this library appealed to me. What alternate library or method should I use to do this as a pre-processing step?

**Comment by polm** [comment]
I've never had to use language detection seriously, so I can't recommend anything in particular, but any library that does just language detection should be fine.

I don't think the feature makes sense as part of a spaCy pipeline because spaCy pipelines are designed to streamline a variety of processing that is assumed to be in the same language. So if you had a site that expects input in English or Spanish, for example, it doesn't make sense to have an English pipeline that tells you the text is in Spanish. I would structure it more like this:

```
es = spacy.load("my_spanish")
en = spacy.load("my_english")

lang = langdetect(input_text)
if lang == "es":
    doc = es(input_text)
elif lang == "en":
    doc = en(input_text)
else:
    # ignore it or something
```

**Comment by benjlis** [comment]
OK, thanks. There doesn't seem to be much available in free language detection libraries, a requirement since this is for academic work and will run against hundreds of thousands of documents.

**Comment by polm** [comment]
I am not sure where you are looking, but there seem to be plenty of open source options? `spacy-langdetect` used [langdetect](https://github.com/Mimino666/langdetect) under the hood, and [lingua](https://github.com/pemistahl/lingua-py) seems to be a more recent library that links to several alternatives.

**Comment by benjlis** [comment]
Thanks very much for these suggestions. Neither of these came up during my Google search.

**Comment by polm** [comment]
For reference, after trying a few things my final Google search query was `site:github.com human language detector python`.

**Comment by polm** [comment]
This will be resolved by #11074.

**Comment by polm** [comment]
#11074 was merged, so this is resolved.
